pvr.h Refactor + Cleanup

gyrovorbis opened the pull request:
Guys, I'm sorry to say, but I have long thought `pvr.h` is probably the single worst source code file in all of KOS for a variety of offenses:
- It's too goddamn big and bloated, it needs to be broken down into multiple files. This makes it extremely hard to navigate through the file or to tell what is relevant when you're just trying to render a scene with the API.
- It is extremely bad about mixing driver implementation details and actual public API. The learning curve on this API is already extremely high, now factor in having to wade through register/memory maps and macros for internal command layouts when you're just trying to figure out how to render a polygon.
- It's very unidiomatic, newbish C: *everything* is an integer. Whether it's a bit mask, an unsigned type, something that is better suited as an enum, or something that's used as a bool, we're doing the pathological worst-case thing and are just making everything ever an int.

How much of the PVR API's learning curve is artificial due to the file just being a monstrosity? Because I honestly think the API itself is quite well designed and intuitive.

I am *positive* that I can clean this up in the same manner that I've been cleaning up other APIs/drivers as I have been touching them and break the thing into logical components. I'm also positive that I can do this all 100% without breaking the API, which I'm making a hard requirement, so don't worry! I will be testing that this still compiles against every PVR-using application I can get my hands on.

List of things this PR entails so far:

1) Dividing pvr.h into a series of individual subheaders within /dc/pvr
  which then get included within the main pvr.h header.
2) Cleaning up the styling and C idiomacy of the entire API
    a. boolean types where applicable
    b. stdint types where applicable
    c. enumeration where applicable
    d. typedefs where applicable
    e. const correctness
    f. struct tag names for all typedefed structs
3) Optimized pvr_poly_cxt_t structure for less memory (196 -> 73 bytes)
    a. packing all related enumerated types
    b. packing outer struct
    c. packing inner anonymous structs
    e. using booleans instead of ints for simple toggle fields
4) Added a todo list
5) Expanded upon documentation
    - Added more detailed descriptions on what primitive contexts are
      and how to use them
        * Added a warning about performance loss of recompiling every frame 
     - Removed some legacy documentation that wasn't relevant anymore 
6) Configured Doxygen so that the first line of comments above a symbol is assumed to be the \brief line without needing the tag to save
space
7) Created pvr_legacy.h file
    - where all macros which are doing nothing but reinventing boolean
      1/0 value types can go to get deprecated and die
    - we're now just making those fields bool

gyrovorbis committed:
Started MAJOR pvr.h refactor + cleanup endeavor.

1) Dividing pvr.h into a series of individual subheaders within /dc/pvr
  which then get included within the main pvr.h header.
2) Cleaning up the styling and C idiomacy of the entire API
    a. boolean types where applicable
    b. stdint types where applicable
    c. enumeration where applicable
    d. typedefs where applicable
    e. const correctness
    f. struct tag names for all typedefed structs
3) Optimized pvr_poly_cxt_t structure for less memory (196 -> 73 bytes)
    a. packing all related enumerated types
    b. packing outer struct
    c. packing inner anonymous structs
    e. using booleans instead of ints for simple toggle fields
4) Added a todo list
5) Expanded upon documentation
    - Added more detailed descriptions on what primitive contexts are
      and how to use them
        * Added a warning about performance loss of recompiling every
	  frame
    - Removed some legacy documentation that wasn't relevant anymore
6) Configured Doxygen so that the first line of comments above a symbol
   is assumed to be the \brief line without needing the tag to save
space
7) Created pvr_legacy.h file
    - where all macros which are doing nothing but reinventing boolean
      1/0 value types can go to get deprecated and die
    - we're now just making those fields bool

gyrovorbis committed:
PVR API working fine after massive refactor!

Saving progress now after refactor, because all examples are building
and working fine. This is how the PVR API will be broken up within
subheaders, I just have to clean each one up now and redo the Doxygen
for them.

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_context.h:
```diff
@@ -0,0 +1,635 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_dma.h:
```diff
@@ -0,0 +1,259 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_fog.h:
```diff
@@ -0,0 +1,122 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_mem.h:
```diff
@@ -0,0 +1,110 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_misc.h:
```diff
@@ -0,0 +1,184 @@
+    \return                 The packed color value
+*/
+#define PVR_PACK_COLOR(a, r, g, b) ( \
+                                     ( ((uint8)( a * 255 ) ) << 24 ) | \
```
```suggestion
                                     ( ((uint8_t)( a * 255 ) ) << 24 ) | \
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_misc.h:
```diff
@@ -0,0 +1,184 @@
+*/
+#define PVR_PACK_COLOR(a, r, g, b) ( \
+                                     ( ((uint8)( a * 255 ) ) << 24 ) | \
+                                     ( ((uint8)( r * 255 ) ) << 16 ) | \
```
```suggestion
                                     ( ((uint8_t)( r * 255 ) ) << 16 ) | \
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_misc.h:
```diff
@@ -0,0 +1,184 @@
+#define PVR_PACK_COLOR(a, r, g, b) ( \
+                                     ( ((uint8)( a * 255 ) ) << 24 ) | \
+                                     ( ((uint8)( r * 255 ) ) << 16 ) | \
+                                     ( ((uint8)( g * 255 ) ) << 8 ) | \
```
```suggestion
                                     ( ((uint8_t)( g * 255 ) ) << 8 ) | \
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_misc.h:
```diff
@@ -0,0 +1,184 @@
+                                     ( ((uint8)( a * 255 ) ) << 24 ) | \
+                                     ( ((uint8)( r * 255 ) ) << 16 ) | \
+                                     ( ((uint8)( g * 255 ) ) << 8 ) | \
+                                     ( ((uint8)( b * 255 ) ) << 0 ) )
```
```suggestion
                                     ( ((uint8_t)( b * 255 ) ) << 0 ) )
```

andressbarajas commented on kernel/arch/dreamcast/hardware/pvr/pvr_prim.c:
```diff
@@ -321,7 +321,7 @@ void pvr_sprite_cxt_txr(pvr_sprite_cxt_t *dst, pvr_list_t list,
-void pvr_sprite_compile(pvr_sprite_hdr_t *dst, pvr_sprite_cxt_t *src) {
+void pvr_sprite_compile(pvr_sprite_hdr_t *dst, const pvr_sprite_cxt_t *src) {
     int u, v;
     uint32 txr_base;
```
```suggestion
    uint32_t txr_base;
```

andressbarajas commented on kernel/arch/dreamcast/hardware/pvr/pvr_prim.c:
```diff
@@ -457,7 +457,7 @@ void pvr_mod_compile(pvr_mod_hdr_t *dst, pvr_list_t list, uint32 mode,
-void pvr_poly_mod_compile(pvr_poly_mod_hdr_t *dst, pvr_poly_cxt_t *src) {
+void pvr_poly_mod_compile(pvr_poly_mod_hdr_t *dst, const pvr_poly_cxt_t *src) {
     int u, v;
     uint32  txr_base;
```
```suggestion
    uint32_t  txr_base;
```

andressbarajas commented on kernel/arch/dreamcast/hardware/pvr/pvr_scene.c:
```diff
@@ -70,7 +70,7 @@ void * pvr_vertbuf_tail(pvr_list_t list) {
 
-void pvr_vertbuf_written(pvr_list_t list, uint32 amt) {
+void pvr_vertbuf_written(pvr_list_t list, size_t amt) {
     uint32 val;
```
```suggestion
    uint32_t val;
```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_texture.h:
```diff
@@ -0,0 +1,136 @@
+void pvr_txr_load_kimg(const kos_img_t *img, pvr_ptr_t dst, pvr_txr_load_t flags);
+
+__END_DECLS
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_palette.h:
```diff
@@ -0,0 +1,93 @@
+
+__END_DECLS 
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_prim.h:
```diff
@@ -0,0 +1,399 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr/pvr_legacy.h:
```diff
@@ -0,0 +1,147 @@
+
+__END_DECLS
+
+#endif
```
```suggestion
#endif

```

andressbarajas commented on kernel/arch/dreamcast/include/dc/pvr.h:
```diff
@@ -20,1718 +21,189 @@
+        - Multipass rendering
+        - pvr_poly_ctx_t::specular2
+            - specular enabled within modifier?
+        - Missing fog mode
```
Oh I see you bundled all the todo on the main header

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Intermediate commit.

gyrovorbis committed:
Final Doxygen Cleanup Pass

- Also found an annoying side-effect of pvr_fog_Table_custom() modifying
  its input LUT table as its being clamped...
- Redid the Doxygen, reconfirmed it works.
- Rebuilt everything, confirmed nothing broke.

gyrovorbis committed:
Apply suggestions from code review

Co-authored-by: Andy Barajas <andress.barajas@gmail.com>

gyrovorbis commented on kernel/arch/dreamcast/include/dc/pvr/pvr_fog.h:
```diff
@@ -0,0 +1,122 @@
+    \ingroup pvr_fog
+
+    This function sets the fog color for vertex fog. 0-1 range for all colors.
+    This function is currently not implemented, as vertex fog is not supported
```
Should already have one at the top of `pvr.h`! But yep!

gyrovorbis committed:
Update kernel/arch/dreamcast/include/dc/pvr/pvr_context.h

Co-authored-by: Andy Barajas <andress.barajas@gmail.com>

gyrovorbis committed:
Update kernel/arch/dreamcast/include/dc/pvr/pvr_prim.h

Co-authored-by: Andy Barajas <andress.barajas@gmail.com>

gyrovorbis commented:
*pant* *cough* *wheeze*

SHE'S DONE!!!! Everything has been rebuilt, tested, redocumented, re-reviewed, etc... I know this is going to be a PITA to look at, but 90% of it was just moving stuff around, with the other 10% being trivial styling like using idiomatic C types.

darcagn committed:
Merge branch 'master' into pvr.h_cleanup

andressbarajas commented on kernel/arch/dreamcast/hardware/pvr/pvr_fog.c:
```diff
@@ -229,10 +229,7 @@ void pvr_fog_table_color(float a, float r, float g, float b) {
-    (void)r;
-    (void)g;
-    (void)b;
+    PVR_SET(PVR_FOG_VERTEX_COLOR, PVR_PACK_COLOR(a, r, g, b));
```
I see that you still have the assert and commented out version of this line below.  Does it work now or is it still not supported?

gyrovorbis replied:
Wow, good eye...

while I was refactoring, this one really bothered me, because I couldn't understand how it could be unsupported, tbh. Maybe this code was from before support for it got added to the polygon/sprite contexts which set the corresponding field in the polygon/sprite header? ...just not seeing why this wouldn't work.

...but since I haven't actually tested this, I'm going to go ahead and remove this line again... I have captured a `\todo` item for this, but if anyone wants to look into it for another PR, it might just magically work with modern KOS.

gyrovorbis commented on kernel/arch/dreamcast/hardware/pvr/pvr_fog.c:
```diff
@@ -229,10 +229,7 @@ void pvr_fog_table_color(float a, float r, float g, float b) {
-    (void)r;
-    (void)g;
-    (void)b;
+    PVR_SET(PVR_FOG_VERTEX_COLOR, PVR_PACK_COLOR(a, r, g, b));
```
```suggestion
```

gyrovorbis committed:
Update kernel/arch/dreamcast/hardware/pvr/pvr_fog.c

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Whitespace OCD + improved Doxygen for pvr_prim.h.

gyrovorbis committed:
Merge branch 'pvr.h_cleanup' of github.com:KallistiOS/KallistiOS into pvr.h_cleanup

gyrovorbis committed:
Filled in more missing \brief file desriptions.

tsymalla commented on doc/Doxyfile:
```diff
@@ -217,7 +217,7 @@ JAVADOC_BANNER         = NO
 # requiring an explicit \brief command for a brief description.)
 # The default value is: NO.
```
Needs update since you changed the default value

darcagn replied:
That's the default value for the upstream example Doxyfle, so we'll keep the text as is here.

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

andressbarajas commented:
No, it has the 2.2.0 tag on it so it will not be in the 2.1.0 release.

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis commented:
Rebased and ready-to-go for `2.2.0`!

gyrovorbis committed:
Updated CHANGELOG.md.

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

pcercuei commented:
I believe you could rename `<pvr/pvr_foo.h>` to `<pvr/foo.h>`.

I also think that "pvr_regs.h" and "pvr_prim.h" contain implementation details that should not be directly visible in the API - at least not by including <kos.h>. Stuff like register addresses etc. should not be part of the visible API.

Then, I think some fields should be renamed (and we can use unions to avoid breaking the API). For instance the `cxt.blend.dst_enable`, documented as "dest blending enable", and usable with `PVR_BLEND_ENABLE`, does actually something very differently (set the second accumulation buffer as the target).

Finally, and this can be part of a different PR, but it would be great to have the geometry header structures directly usable, without having to use a "context" structure that is "compiled" later.

gyrovorbis commented:
> I believe you could rename `<pvr/pvr_foo.h>` to `<pvr/foo.h>`.
> 
> I also think that "pvr_regs.h" and "pvr_prim.h" contain implementation details that should not be directly visible in the API - at least not by including <kos.h>. Stuff like register addresses etc. should not be part of the visible API.
> 
> Then, I think some fields should be renamed (and we can use unions to avoid breaking the API). For instance the `cxt.blend.dst_enable`, documented as "dest blending enable", and usable with `PVR_BLEND_ENABLE`, does actually something very differently (set the second accumulation buffer as the target).
> 
> Finally, and this can be part of a different PR, but it would be great to have the geometry header structures directly usable, without having to use a "context" structure that is "compiled" later.

I 100% agree with you on every front. I simply cut this PR off at not doing anything at all to break the API or add any new features, because it was already becoming a beast.

I don't know if those things should be a part of this PR or not, but agreed, we should totally do it for 2.2.0!

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

gyrovorbis committed:
Update pvr_misc.h

Fixed bug in `PVR_PACK_COLOR()` macro expansion which didn't make its way into this branch.

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

andressbarajas commented:
I have verified that these changes do not break the gldc or pvr examples.

andressbarajas commented:
^ These two PRs should fix all compile errors for the examples after this PR has been merged.

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

andressbarajas commented:
@QuzarDC Anything we can do to help push this along to being merged?

gyrovorbis committed:
Merge branch 'master' into pvr.h_cleanup

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr.h:
```diff
@@ -17,1773 +18,206 @@
-    This function is currently not implemented, as vertex fog is not supported
-    by KOS. Calling this function will cause an assertion failure.
+    \todo 
+        - Auxiliary accumulation buffer
```
That is already supported.

gyrovorbis replied:
You're right. At the time I had no idea how it was actually used, and everyone told me it wasn't something they knew how to use either... just need probably better docs and/or example.

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr.h:
```diff
@@ -17,1773 +18,206 @@
+    PVR_LIST_TR_POLY = 2,  /**< Translucent polygon list */
+    PVR_LIST_TR_MOD  = 3,  /**< Translucent modifier list*/
+    PVR_LIST_PT_POLY = 4,  /**< Punch-thru polygon list */
+    PVR_LIST_COUNT   = 5   /**< Number of list types */
```
No need for the explicit numbering. The same goes for every other enum.

gyrovorbis replied:
You're right, but... now that I looked back, I remembered why I did it... Because their values actually do matter when the PVR itself is expecting them to be equal to certain things, and the values are going directly to the HW. I was worried not making them explicit would make it look like it's safe to change the values while explicitly enumerating would make it more clear that they are fixed?

...but I'm not very strong on this opinion. I actually dislike explicit enumeration like this in general. If you don't think this is an exception that warrants it, then I'm going with your vote, because I don't feel strongly either way.

pcercuei replied:
The way I see it... those enums represent hardware values and as such shouldn't be changed, but:
- if for some reason we need new values we'd add them to the end, so it wouldn't be a problem;
- a PR that casually re-orders the enum entries would most likely cause breakages and wouldn't pass peer review (not my review at least).

gyrovorbis replied:
Done.

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr.h:
```diff
@@ -1819,29 +253,32 @@ void pvr_mem_stats(void);
     
-    \return                 Non-zero if vertex DMA was enabled at init time
+    \retval  true       Vertex DMA was enabled at init time
+    \retval  false      Vertex DMA was not enabled at init time
```
Maybe `\return True if vertex DMA was enabled at init time, false otherwise`

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr/pvr_context.h:
```diff
@@ -0,0 +1,612 @@
+    PVR_FILTER_BILINEAR   = 2,   /**< \brief Bilinear interpolation */
+    PVR_FILTER_TRILINEAR1 = 4,   /**< \brief Trilinear interpolation pass 1 */
+    PVR_FILTER_TRILINEAR2 = 6    /**< \brief Trilinear interpolation pass 2 */
+} pvr_filter_t;
```
I would actually use 0/1/2/3 here, because it is clear that bit 0 does not encode for the filter; and fixing it in pvr_poly_compile() and friends. That shouldn't break the API.

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr/pvr_fog.h:
```diff
@@ -0,0 +1,134 @@
+
+    \sa pvr_fog_table_exp2(), pvr_fog_table_exp(), pvr_fog_table_linear()
+*/
+void pvr_fog_table_custom(const float tbl1[]);
```
Either:
`const float *table`
Or:
`const float table[129]`

pcercuei commented on kernel/arch/dreamcast/include/dc/pvr/pvr_prim.h:
```diff
@@ -0,0 +1,359 @@
+    these from your pvr_poly_cxt_t and use it for submission to the hardware.
+*/
+typedef struct pvr_poly_hdr {
+alignas(32)
```
`alignas(32)` is not... aligned?

gyrovorbis replied:
I don't know why, in my head it looked better like that, but now that you mention it, it actually looks like shit to me, so... Reverting. lol.

pcercuei reviewed (changes requested):
A few things I noted too, is that some doxygen stuff (\defgroup, \see etc.) was dropped between the main pvr.h and the new headers.

gyrovorbis committed:
Addressing review feedback.

gyrovorbis committed:
Addressed review feedback.

QuzarDC reviewed (commented):
Sorry for the lateness of these. I even wrote them a month ago when I was working on #874 but just hadn't hit commit. Almost all of them relate to the pieces I left out.

pcercuei commented on kernel/arch/dreamcast/hardware/pvr/pvr_fog.c:
```diff
@@ -144,10 +144,9 @@ union ieee32_t {
-#define NEG(n) ( (n)>(0.0f) ? (-n):(n) )
-#define CLAMP01(x) \
-    ( (x)<(0.0f) ? ((x) = (0.0f)) : ((x)>(1.0f) ? ((x)=(1.0f)) : (x)) )
+#define ABS(n)     ((n) < (0.0f) ? (-n):(n))
+#define NEG(n)     ((n) > (0.0f) ? (-n):(n))
+#define CLAMP01(x) ((x) < (0.0f) ? ((0.0f)) : ((x) > (1.0f) ? ((1.0f)) : (x)))
```
One thing I would add: unless you need to write something that needs macro constructs, please always use static inline functions instead.

QuzarDC commented:
Worked through a cleaning up and splitting off a majority of the changes here in #874 #934 #1051 and #1288 . There's still quite a bit with our current `pvr.h` being double the size of the one here, but it's still half the size as it was when it started.

I've added the new label 'Revisit' to keep this searchable with other closed PRs that have useful changes, but it doesn't seem like there's any value to holding it open as a draft any further work would certainly be a new PR and just reference this.
